text/0715-construct-upgrade-support.md

```diff
@@ -0,0 +1,1251 @@
+
+This process is cumbersome and typically requires administrative permissions to modify the CI system. Given
+these challenges, we recommend using Ops-led deployment for teams with complex pipeline setups until a more
+elegant solution is developed in the future updates.
+
+## Rollback
+
+After performing the execution step in the stack, the CLI will proceed with the deployment
+(assuming that is your choice). If the deployment fails, and CloudFormation rolls it back which
+will bring the resources back to their original states.
+
+## Programmatic access
+
+The same migration feature is also available in the CDK toolkit library:
+
+```typescript
+declare const toolkit: Toolkit;
+declare const cx: ICloudAssemblySource;
+
+// Or, if you just want to cosntruct upgrade the stacks:
+await toolkit.constructUpgrade(cx);
+```
+
+## Limitations and failure modes
+
+### Unintended Construct Upgrades: Mitigating Risks for Unaware Users
+
+There may be users who are unaware of this officially supported construct upgrade CLI command. They may simply
+attempt to upgrade their CDK code from V1 construct to V2 construct and run cdk deploy directly. This would most
+certainly cause resource replacement, data loss, or service downtime. CDK CLI currently does not have a mechanism
+to detect that users are attempting a construct upgrade and prevent potentially destructive changes.
+
+A future enhancement would add detection capabilities to the `cdk deploy` command. The CLI would analyze the change
+set prior to deployment and compare it against the context file mappings (See [Appendix C: Main Resources](#appendix-c-main-resources-computation)):
+
+```json
+{
+  "aws-cdk-lib.aws_dynamodb.TableV2": {
+    "strategy": "Import",
+    "source": ["AWS::DynamoDB", "Custom::ReplicationDeletion"],
+    "target": ["AWS::DynamoDB::GlobalTable"]
+  }
+}
+```
+
+If the CLI detects resource removals matching the source types and resource additions matching the target types, it would either:
+
+- Display a warning message about the potential construct upgrade scenario and prompt users to use `cdk construct-upgrade`
+  and the corresponding migration guide, or
+- Present a confirmation prompt requiring explicit user approval to proceed (unless `--require-approval never` is specified).
+
+This enhancement would help prevent accidental resource replacements by guiding users toward the proper construct upgrade process.
+
+### Limitation: In-Place Migration Strategy
+
+This strategy utilizes the CDK/CloudFormation refactoring feature and inherits its limitations. It experiences
+the [same limitations](https://github.com/aws/aws-cdk-rfcs/blob/otaviom/refactoring-support/text/0162-refactoring-support.md#pipelines-with-version-superseding)
+as described in the CDK Refactoring RFC. Primarily, these limitations affect pipelines with version superseding,
+potentially causing issues in complex deployment scenarios.
```

Comment:
Would be great to understand the complex deployment scenarios and how likely they are to cause issues.

Reply:
The phrasing is inaccurate, it's actually talking about the issue right above where the pipeline can't customize the CLI command to run.